packages/interactivity/src/proxies/state.ts

```diff
@@ -0,0 +1,200 @@
+/**
+ * External dependencies
+ */
+import { signal, type Signal } from '@preact/signals';
+
+/**
+ * Internal dependencies
+ */
+import { createProxy, getProxy, shouldProxy } from './registry';
+import { PropSignal } from './signals';
+import { setNamespace, resetNamespace } from '../hooks';
+
+type WellKnownSymbols =
+	| 'asyncIterator'
+	| 'hasInstance'
+	| 'isConcatSpreadable'
+	| 'iterator'
+	| 'match'
+	| 'matchAll'
+	| 'replace'
+	| 'search'
+	| 'species'
+	| 'split'
+	| 'toPrimitive'
+	| 'toStringTag'
+	| 'unscopables';
+
+const proxyToProps: WeakMap<
+	object,
+	Map< string | symbol, PropSignal >
+> = new WeakMap();
+
+const objToIterable = new WeakMap< object, Signal< number > >();
+
+const wellKnownSymbols = new Set(
+	Object.getOwnPropertyNames( Symbol )
+		.map( ( key ) => Symbol[ key as WellKnownSymbols ] )
+		.filter( ( value ) => typeof value === 'symbol' )
+);
+
+const descriptor = Object.getOwnPropertyDescriptor;
+
+let peeking = false;
+
+const stateHandlers: ProxyHandler< object > = {
+	get( target: object, key: string, receiver: object ): any {
+		const desc = descriptor( target, key );
+
+		/*
+		 * If peeking, the property comes from the Object prototype, or the key
+		 * is a well-known symbol, then it should not be processed.
+		 */
+		if (
+			peeking ||
```

Comment:
I might be misunderstanding something, but is this not a potential race condition? 

Can it happen that `peeking === true` because some key `A` is being peeked at, but at the very same time, the `get()` handler is invoked for some **different** key `B`?

Reply:
Gotcha, thanks guys!